### ISSUE TITLE
HTML manipulation allows bypass payment restrictions

### ISSUE THREAD

**Issue by pszostok** [PrestaShop/PrestaShop#36619]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md).
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add attachments

This security vulnerability was reported 29th of January, but since then there is no feedback about fixing this issue, so I decided to make it public, so everyone can protect their shops by them self.

We have identified a significant vulnerability in PrestaShop that allows users to bypass restrictions and select a payment method that should not be available for a given shipping method.

**Potential source of the bug:**
The issue likely resides within the PaymentModule class, specifically within the validateOrder function. It appears this function does not adequately validate the chosen payment method against the available options for the selected shipping method. Introducing a check at this stage to match payment methods with shipping options and throwing an exception if there's a mismatch could potentially address this vulnerability.

**Why is this critical?**
Some stores automatically generate shipping labels, and for certain shipping methods, payment upon package receipt (cash on delivery) is allowed. If a user selects a standard shipping method (requiring prepayment) but manipulates the HTML to trick PrestaShop into processing the order as cash on delivery, the store might dispatch the package without securing payment. If the discrepancy goes unnoticed by the store administrator, it could lead to significant losses.

### Steps to reproduce

1. Create two shipping methods and two payment methods.
2. Configure the system so the first shipping method only allows the first payment method, and the second shipping method allows only the second payment method.
3. Place a new order and proceed to the final step (payment method). Inspect the DOM for the form inside the div with the id "pay-with-payment-option-1-form". Modify the form's action URL from "/module/ps_cashondelivery/validation" to another payment method's URL, for example, "/module/ps_wirepayment/validation".
4. Submit the order.

### PrestaShop version(s) where the bug happened

1.7.7.4, 8.1.7

**Comment by matks** [comment]
Hi @pszostok I'm the one who answered first to your message 😉 

I know you consider this a security vulnerability but from my point of view this has nothing to do with security. There is no data stolen, there is no intrusion. This does not match the criteria for a vulnerability.

However I agree this is a functional bug: one level of validation could be added to prevent the scenario you mention, and this could be improved. This is a limitation of the current state of the system.

**Comment by pszostok** [comment]
@matks Thanks for your response.

I understand that this is not a data-leak vulnerability, so I extended the basic 90-day rule, which is typically applied to security vulnerabilities, to nearly 6 months. Nevertheless, in my opinion, this bug is still a critical issue. If left unfixed, it could result in significant losses for many shops worldwide.

Delaying the fix allows more people to become aware of this exploit, potentially using it for their own profit. We have already observed a few instances where someone has exploited this bug.

**Comment by ShaiMagal** [comment]
I aggree with @matks, it isnt security vulnerability. And calling this "exploit" is little strange...

Its "only" bug, not security vulnerability.
In every case, its not critical...

I dont believe normal users will "exploit" this and professionals dont care about this type of "vulnerability"...

Anyway, it should be fixed, if there is possibility to fix it easily. (One more level of checking at validation controller of payment modules - ideally globally for whole system, not single module separately)

**Comment by matks** [comment]
> In every case, its not critical...

Yes and no.

To be exploited, the bug needs that you have 2 payment methods available on your website, one that needs you to pay NOW and one that allows you to pay later. In some countries like Germany, "pay when shipment has been received" is quite common.

So it's hard to label this as critical, but it's not trivial either

**Comment by ShaiMagal** [comment]
I understand what you mean, cash on delivery - it's common in Czech republic as well.

But, normal customer don't care about "exploiting" it and attacker can be 1 of 100000 customers.

For me it's not critical or security issue.

**But I agree, if this is possibility how to fix, it will be good to do this :)**

**Comment by ArniPL** [comment]
Of course. Regardless of the bug exposure or popularity, this lack of validation should be fixed and the [PR is in the works right now](https://github.com/PrestaShop/PrestaShop/pull/36648). 🙂

**Comment by florine2623** [comment]
Hello @pszostok ,

I reproduce the issue with PrestaShop version 8.1.x, I'll add this to the backlog so it can be fixed.

https://github.com/user-attachments/assets/4e27608e-ae9f-4a8e-aed8-645a069a8fc2

Please be aware that some issues might take a very long time to be resolved. If this one is important to you and you cannot wait for it to be fixed on the project’s own time, we strongly suggest you consider [contacting a professional](https://www.prestashop-project.org/support/) to help you. 

If you fix the issue on your end, please [contribute it back to the project](https://devdocs.prestashop.com/1.7/contribute/contribute-pull-requests/). Remember that the more people contribute, the better PrestaShop becomes for everyone.

Thank you

**Comment by boo-code** [comment]
Still reproduces on 9.2.x, and the mechanism is narrow enough to state precisely.

The carrier restriction is applied when core decides which modules receive the payment hooks, in `Hook::getHookModuleExecList()`, which adds a `module_carrier` condition for `displayPayment`, `displayPaymentEU` and `paymentOptions`. Nothing in the order creation path reads that table: in `classes/PaymentModule.php` it is only written, by `install()`, `uninstall()` and `addCheckboxCarrierRestrictionsForModule()`, and neither `validateOrder()` nor `createOrderFromCart()` consults it. So the restriction governs what is rendered, not what is accepted.

Measured on a 9.2 shop, toggling the restriction row for `ps_checkpayment` on the cart's carrier and asking core which modules are hooked on `paymentOptions`:

```
cart=15 carrier=2 ref=2
module_carrier row exists = true
offered for this carrier (before)              = YES
offered after removing the restriction row     = no
restored, offered                              = YES
```

The module disappears from the offered list exactly as configured, which is why the submitted value is the only thing left unchecked.

One constraint worth recording before anyone adds the check to `validateOrder()`, since that is the natural place and #36648 goes there. That method is not front office only. `src/Adapter/Order/CommandHandler/AddOrderFromBackOfficeHandler.php` calls it to create an order from the back office, where a merchant records a payment for a carrier the restriction was never meant to govern, and `classes/order/Order.php` calls it as well. A guard placed there without distinguishing the origin would reject legitimate back office orders. The check has to apply to the customer driven path only.

#36648 targets `8.2.x` and has been waiting for author since 2025-01-20, so it would not reach 9.x even if it were finished.
